### PR TITLE
Fix a bug where "connection: close" header is stripped  in a HTTP/1.1 response

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

`connection: close` header that should be preserved is removed while an Armeria server response is converted to a Netty HTTP/1 response. We allowed users to send `connection: close` #4531. The connection was correctly closed after fully writing the response but `connection: close` wasn't sent correctly.

Modifications:

- Make `KeepAliverHandler` closed when `connection: close` header is specified in the response headers.

Result:

In HTTP/1.1, if `connection: close` is set in response headers, the value is correctly sent to clients.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/ServerHttp1ObjectEncoder.java [hunk 1]
```diff
@@ -111,11 +111,8 @@ private HttpResponse convertHeaders(ResponseHeaders headers, boolean endStream,
 
     private void convertHeaders(HttpHeaders inHeaders, io.netty.handler.codec.http.HttpHeaders outHeaders,
                                 boolean isTrailersEmpty) {
-        if (keepAliveHandler.needsDisconnection()) {
-            sentConnectionCloseHeader = true;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/ServerHttp1ObjectEncoder.java:115]: Could you remove the field `sentConnectionCloseHeader`?

**Review by minwoox — APPROVED** [review]
Thanks!

File: core/src/test/java/com/linecorp/armeria/server/Http1ServerKeepAliveTest.java [hunk 2]
```diff
@@ -91,4 +111,36 @@ void noKeepAlive(ExchangeType exchangeType) throws Exception {
...
+            writer.print("\r\n");
+            writer.flush();
+
+            final BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            assertThat(in.readLine()).isEqualTo("HTTP/1.1 200 OK");
+
+            String line;
+            boolean hasConnectionClose = false;
+            while ((line = in.readLine()) != null) {
+                System.out.println("### " + line);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/Http1ServerKeepAliveTest.java:131]: Maybe use `logger`?

Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/server/Http1ServerKeepAliveTest.java:131]: It is cruft. 😓

File: core/src/main/java/com/linecorp/armeria/server/AbstractHttpResponseHandler.java [hunk 3]
```diff
@@ -75,6 +75,7 @@ boolean isDone() {
 
     void disconnectWhenFinished() {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/AbstractHttpResponseHandler.java:76]: Would it make sense to remove this method and call `responseEncoder.keepAliveHandler().disconnectWhenFinished();` directly?
We can probably integrate the closing logic in `HttpServerHandler` with this in one place.

**Review by jrhee17 — APPROVED** [review]
Thanks for the fix @ikhoon ! 🙇 👍 🙇

**Review by jrhee17 — APPROVED** [review]
Thanks for the fix @ikhoon ! 🙇 👍 🙇

File: core/src/main/java/com/linecorp/armeria/server/AbstractHttpResponseHandler.java (hunk 3 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/AbstractHttpResponseHandler.java:76]: Let me clean up the closing logic in the following work, perhaps, after merging `WebSocketService` PR.

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/AbstractHttpResponseHandler.java:76]: Thanks a lot! 🙇

**Comment by ikhoon** [comment]
Resolved.

**Review by trustin — APPROVED** [review]
:bow:

**Pull request merged**